### ISSUE TITLE
Build failure

### ISSUE THREAD

**Issue by fangyi-zhou** [FStarLang/kremlin#169]
```
ocamlbuild -I src -I lib -I parser -I kremlib -use-menhir -use-ocamlfind -classic-display -menhir "menhir --infer --explain" Kremlin.native
ocamlfind ocamlc -c -g -annot -bin-annot -strict-sequence -w @1-2-3@8..12@14..21@23..29-30@31..38-39-40-41@43@57 -package ppx_deriving.std -package ppx_deriving_yojson -package zarith -package pprint -package unix -package menhirLib -package sedlex -package process -package fix -package wasm -package visitors.ppx -package visitors.runtime -I src -I parser -I kremlib -I lib -o src/CFlatToWasm.cmo src/CFlatToWasm.ml
+ ocamlfind ocamlc -c -g -annot -bin-annot -strict-sequence -w @1-2-3@8..12@14..21@23..29-30@31..38-39-40-41@43@57 -package ppx_deriving.std -package ppx_deriving_yojson -package zarith -package pprint -package
unix -package menhirLib -package sedlex -package process -package fix -package wasm -package visitors.ppx -package visitors.runtime -I src -I parser -I kremlib -I lib -o src/CFlatToWasm.cmo src/CFlatToWasm.ml
File "src/CFlatToWasm.ml", line 196, characters 18-32:
196 |   [ dummy_phrase (W.Ast.TeeLocal (mk_var (env.n_args + 2)));
                        ^^^^^^^^^^^^^^
Error: Unbound constructor W.Ast.TeeLocal
Command exited with code 2.
Makefile:22: recipe for target 'minimal' failed
make: *** [minimal] Error 10
```

I am using wasm 1.1, maybe that's the problem?

**Comment by msprotz** [comment]
ah, I was unaware that the wasm package had been updated -- do you know what replaced the TeeLocal instruction?

**Comment by fangyi-zhou** [comment]
Is Kremlin supposed to support OCaml >= 4.08.0?
The opam package wasm 1.0 supports < 4.08.0, wasm 1.1 supports >= 4.08.0.
I'm using 4.08.1 so I encountered the issue.

Therefore I don't think this would be an easy fix... Maybe it's necessary to contact wasm maintainer to relax the version requirements

**Comment by msprotz** [comment]
Unfortunately, for various reasons in Everest, we can only work with at most OCaml 4.06.1... so this error flew under our radar. 

I think we should just be explicit that Everest doesn't work with recent OCaml versions. Once we fix the sedlex replacement in FStar, we can consider an upgrade and then I'll fix kremlin. Does that sound good?

**Comment by fangyi-zhou** [comment]
I see, I can easily create a switch for 4.06.1 and build kremlin.
Sedlex replacement in F* is a bit tricky, I tried to create a patch for that and didn't have time to finish.
But I've been able to use F* with 4.08.1 with my current local configuration

**Comment by paul-snively** [comment]
> But I've been able to use F* with 4.08.1 with my current local configuration

It would be very nice to learn how to reproduce this. It's quite problematic to me to hear that F*/Kremlin are so far behind current OCaml.

**Comment by fangyi-zhou** [comment]
> > But I've been able to use F* with 4.08.1 with my current local configuration
> 
> It would be very nice to learn how to reproduce this. It's quite problematic to me to hear that F*/Kremlin are so far behind current OCaml.

I think you can install required packages using opam, and then build from source? I didn't encounter extra issues with 4.08.1 (F* and Kremlin). If you have specific errors I might be able to help

**Comment by paul-snively** [comment]
I get the same error you described earlier, with wasm 1.1 and OCaml 4.08.1 when I try to build Kremlin.

I'm honestly a bit surprised there doesn't seem to be an effort to have the OCaml build tooling around F*/Kremlin specify dependency versions, when OPAM is one of the best package managers around at actually dealing with that correctly. Is this some deficiency in ocamlbuild? Would transitioning to dune help?

**Comment by msprotz** [comment]
I have unfortunately very few cycles to figure out how to state that kremlin only works with wasm 1.0. Do you know how to achieve that with Dune? I have expressed in another issue that I'm very much in favor of switching to Dune so if you know how to do the migration *and* solve the dependency issue, I'd be more than happy to review and merge any pull request.

Thanks,

Jonathan

**Comment by fangyi-zhou** [comment]
Oh sorry for the confusion. I meant 4.07.1. My apologies for all

**Comment by fangyi-zhou** [comment]
Workaround for OCaml 4.08.1:
- Clone wasm repo, checkout v1.0 tag
- In Makefile, silence warning 3 (for deprecations) by adding "-3" to warn-error
- `make install` in wasm
- build Kremlin as usual

**Comment by paul-snively** [comment]
@fangyi-zhou: Thanks for the workaround tip!

@msprotz: Oh, interesting. Do you happen to know what issue I should look at for the Dune effort?

**Comment by paul-snively** [comment]
OK, I've looked at the "Switch to Dune" ticket and poked around a bit at OCamlbuild, which had always put me off, and I have essentially three thoughts:

1. An OCamlbuild to Dune migration would not be a trivial effort.
2. It's astonishing to me that OCamlbuild does not appear to support dependency version constraints. At all.
3. Even when dependency version constraints are supported (e.g. by Dune), they are Findlib dependency version constraints, and apparently the version field in Findlib META files is optional. Many libraries don't provide a version. Including wasm.

This is sufficiently ridiculous as to cause me to reconsider my interest in the OCaml ecosystem.

**Comment by msprotz** [comment]
I don't think switching to dune would be a huge effort: the build of kremlin itself is simple, and the extraction and linking of libkremlib + tests are handled separately with their own individual Makefiles. So this should be pretty textbook.

Now if Dune doesn't allow a way to require a downgrade of a package, that's more of an issue. But reading from their description:

> It also supports multi-context builds, such as building against several opam roots/switches simultaneously. This helps maintaining packages across several versions of OCaml and gives cross-compilation for free.

So surely that must be achievable. I should also mention that we're going to have to require everyone in everest to upgrade their opam setups owing to a dependency on bigstring, so now might be the right time to flip the switch for requiring dune to build kremlin, too. I just lack the cycles to do this myself.

**Comment by fangyi-zhou** [comment]
FYI: wasm [1.0.1](https://opam.ocaml.org/packages/wasm/wasm.1.0.1/) is released with the version constraint removed.

**Comment by msprotz** [comment]
Thanks very much. This means that KreMLin will be broken at the next OPAM upgrade... do you know what and how we can fix KreMLin to be compatible with the upcoming version of the Wasm package?

ping @denismerigoux who's working on this and may be impacted

**Comment by fangyi-zhou** [comment]
@msprotz wasm 1.0.1 is wasm 1.0 without the version constraint of OCaml.
If KreMLin is not updating to use wasm 1.1, staying on wasm 1.0.1 allows KreMLin to compile with latest OCaml, fixing the initial compilation problem in the first message of this issue.

**Comment by phase** [comment]
I just wanted to confirm that kremlin builds with OCaml `4.09.1`, OPAM `2.0.4`, & wasm `1.0.1`! This should probably be specified in the README but I believe the everest script would also need to be updated.

**Comment by msprotz** [comment]
Wonderful, can you mention this in the README?

**Comment by beurdouche** [comment]
@protz I have just hit that bug, with `ocaml 4.12.0` and `wasm 1.1.1`.

**Comment by msprotz** [comment]
you can merge branch wasm1.1 of the repository into your local master to get unblocked -- I'm holding off on merging that branch until F* moves to a recent OCaml and it becomes official

**Comment by beurdouche** [comment]
Apparently I get an error when installing `ctypes-foreign` with `ocaml 4.12.0`...
```
❯ opam install ctypes-foreign
The following actions will be performed:
  ∗ install   conf-pkg-config 2      [required by ctypes-foreign]
  ∗ install   conf-libffi     2.0.0  [required by ctypes-foreign]
  ∗ install   ctypes-foreign  0.18.0
  ↻ recompile ctypes          0.18.0 [uses ctypes-foreign]
===== ∗ 3   ↻ 1 =====
Do you want to continue? [Y/n] y

<><> Gathering sources ><><><><><><><><><><><><><><><><><><><><><><><><><><>  🐫 
[ctypes.0.18.0] found in cache

<><> Processing actions <><><><><><><><><><><><><><><><><><><><><><><><><><>  🐫 
[ERROR] The compilation of conf-pkg-config failed at "/Users/bbeurdouche/.opam/opam-init/hooks/sandbox.sh build pkg-config --help".

#=== ERROR while compiling conf-pkg-config.2 ==================================#
# context     2.0.8 | macos/arm64 | ocaml-base-compiler.4.12.0 | https://opam.ocaml.org#15d6aa57
# path        ~/.opam/default/.opam-switch/build/conf-pkg-config.2
# command     ~/.opam/opam-init/hooks/sandbox.sh build pkg-config --help
# exit-code   71
# env-file    ~/.opam/log/conf-pkg-config-74001-ffb3fd.env
# output-file ~/.opam/log/conf-pkg-config-74001-ffb3fd.out
### output ###
# sandbox-exec: execvp() of 'pkg-config' failed: No such file or directory



<><> Error report <><><><><><><><><><><><><><><><><><><><><><><><><><><><><>  🐫 
┌─ The following actions failed
│ λ build conf-pkg-config 2
└─ 
╶─ No changes have been performed

The packages you requested declare the following system dependencies. Please make sure they are installed before retrying:
    pkg-config

❯ gmake -j
OCAML_ERROR_STYLE="short" ocamlbuild -I src -I lib -I parser -I kremlib -use-menhir -use-ocamlfind -classic-display -menhir "menhir --infer --explain" Kremlin.native
ocamlfind ocamlopt -c -g -annot -bin-annot -strict-sequence -package compiler-libs -package compiler-libs.common -package ocaml-migrate-parsetree -w @1-2-3@8..12@14..21@23..29-30@31..38-39-40-41@43@57 -package ppx_deriving.std -package ppx_deriving_yojson -package zarith -package pprint -package unix -package menhirLib -package sedlex -package process -package fix -package wasm -package visitors.ppx -package visitors.runtime -I src -I lib -I parser -I kremlib -o src/GenCtypes.cmx src/GenCtypes.ml
+ ocamlfind ocamlopt -c -g -annot -bin-annot -strict-sequence -package compiler-libs -package compiler-libs.common -package ocaml-migrate-parsetree -w @1-2-3@8..12@14..21@23..29-30@31..38-39-40-41@43@57 -package ppx_deriving.std -package ppx_deriving_yojson -package zarith -package pprint -package unix -package menhirLib -package sedlex -package process -package fix -package wasm -package visitors.ppx -package visitors.runtime -I src -I lib -I parser -I kremlib -o src/GenCtypes.cmx src/GenCtypes.ml
findlib: [WARNING] Interface topdirs.cmi occurs in several directories: /Users/bbeurdouche/.opam/default/lib/ocaml, /Users/bbeurdouche/.opam/default/lib/ocaml/compiler-libs
File "src/GenCtypes.ml", line 25, characters 16-32:
Error: Unbound module Versions
Command exited with code 2.
gmake: *** [Makefile:25: minimal] Error 10
```

**Comment by beurdouche** [comment]
Let me move this to another bug, it doesn't seem specific to the `wasm-1.1` branch...

**Comment by fangyi-zhou** [comment]
> The packages you requested declare the following system dependencies. Please make sure they are installed before retrying:
    pkg-config
